command-expansion-server/src/app.ts

```diff
@@ -68,8 +75,38 @@ app.get("/expansion/:expansionId(\\d+)", async (req, res) => {
   return;
 });
 
-app.put("/expansion-set", async (req, res) => {
-  res.status(501).send("PUT /expansion-set: Not implemented");
+app.post("/put-expansion-set", async (req, res) => {
+  const { commandDictionaryId, missionModelId, expansionIds } = req.body.input;
+
+  if (!expansionConfigValidator(req.body.input)) {
+    res.status(400).json({
+      message: `PUT /expansion-set: Invalid request body: ${JSON.stringify(req.body.input)}\n${util.formatWithOptions(
+        { depth: Infinity },
+        expansionConfigValidator.errors
+      )}`,
+    });
+    return;
+  }
+
+  const { rows } = await db.query(`
+    WITH expansion_set_id AS (
+      INSERT INTO expansion_set (command_dict_id, mission_model_id)
+      VALUES (${commandDictionaryId}, ${missionModelId})
+      RETURNING id
+    )
+    INSERT INTO expansion_set_to_rule (set_id, rule_id) VALUES
+      ${expansionIds
+        .map((expansionId: number) => `((SELECT id FROM expansion_set_id), ${expansionId})`)
+        .join(",\n      ")}
+    RETURNING (SELECT id FROM expansion_set_id);
+  `);
```

Comment:
I marked the CodeQL failure as a false positive here since I believe `db.query()` performs appropriate sanitization.

Reply:
I'm all for someone figuring out how to do this appropriately without the string templating, but I just didn't know how to effectively do it

Reply:
Well I think what you have is okay since `Pool.query()` handles the sanitization. I could be very mistaken though so take my conclusion with a grain of salt 😂

Reply:
https://node-postgres.com/features/queries#parameterized-query

Reply:
Let's use parameterized queries as a general rule of thumb.

Reply:
Not necessarily advocating for this approach, but there does seem to be an option that requires no string interpolation at all (even for the $nums):

https://github.com/brianc/node-postgres/issues/957#issuecomment-295583050

Reply:
@Twisol @patkenneally with that in mind, we need to do input validation on the apis as we can't rely on graphologist for it, so let's add an issue for that. @mattdailis thats exactly the kind of incantation we need and should implement. 

I mostly but don't entirely agree. I would think that hasura is less of a service than an api layer that necessarily is coupling to our services. I would consider each of the services uncoupled from each other, but the API layer I would expect is something we can rely on. We change the form of our service endpoints to suit it after all.

Are we replicating the input validation on the rest of our services?

Reply:
Took me a bit, but I found an SQL ritual to summon the correct rows

Reply:
Nice! That looks *clean*.

Reply:
Clean and very complicated to understand lol - the parameters for the array_fill and array_length functions are...weird